server/models/commands/ChangeDataVersionCommand.py

```diff
@@ -20,17 +30,66 @@ def backward_impl(self):
         self.wf_module.save()
         self.wf_module.set_fetched_data_version(self.old_version)
 
+    async def schedule_execute(self) -> None:
+        """
+        Tell renderers to render the new workflow, _maybe_.
+
+        ChangeDataVersionCommand is often created from a fetch, and fetches
+        are often invoked by cron. These tend to be our most resource-heavy
+        operations: e.g., Twitter-accumulate with 1M records. So let's use lazy
+        rendering.
+
+        From our point of view:
+
+            * If self.workflow has notifications, render.
+            * If anybody is viewing self.workflow right now, render.
+
+        Of course, it's impossible for us to know whether anybody is viewing
+        self.workflow. So we _broadcast_ to them and ask _them_ to request a
+        render if they're listening. This gives N render requests (one per
+        Websockets communicator) instead of 1, but we don't mind because
+        spurious render requests are no-ops.
+
+        From the user's point of view:
+
+            * If I'm viewing self.workflow, changing data versions causes a
+              render. (There isn't even any HTTP traffic: the communicator
+              does the work.)
+            * Otherwise, the next time I browse to the page, the page-load
+              will request a render.
+
+        Assumptions:
+
+            * Websockets communicators queue a render when we ask them.
+            * The Django page-load view queues a render when needed.
+        """
+        if await _workflow_has_notifications(self.workflow):
+            await super().schedule_execute()
+        else:
+            await websockets.queue_render_if_listening(
+                self.workflow.id,
+                self.workflow.last_delta_id
+            )
+
+    @classmethod
+    def amend_create_kwargs(cls, *, wf_module, **kwargs):
```

Comment:
What is the purpose of putting this here rather than passing it in from `create()` as before?

Reply:
Deltas are a minefield of races because Websockets is async and database operations are sync. (There's no way to hold a database lock from an async method.) So Delta application kinda goes like this:

1. Call `create()` (async method -- database access not allowed)
2. `create()` calls `create_impl()` (mostly for historical race-culling reasons; for now the one remaining feature is that commands can give `create()` positional arguments, and `create_impl()` uses kwargs like a normal Django model.)
3. `create_impl()` _locks the workflow_ (we're going sync now) and:
    a. Calls `amend_create_args()` to gather pre-Delta state -- this is the only safe spot to query the database
    b. Calls the _actual_ `.objects.create()`
    c. Calls `forward_impl()` (the command-specific instructions to commit the Delta)
    d. Sets `workflow.last_delta`
    e. Calls `.load_ws_data()`, which queries the database to gather the data we'll be sending over WebSockets
4. Notify WebSockets (async method -- database access not allowed)
5. Schedule a render over RabbitMQ (async method -- database access not allowed)

Long story short: querying the database in `create()` connects to the database from an executor thread and never releases the connection. It's a leak. (The max number of leaked connections is the size of the executor thread pool). Come to think of it, we might have a fair number of such bugs.

There's already a fair amount of documentation to this effect in `Delta`; and in the `tabs` branch things are even cleaner.